src/Particular.Analyzers/Cancellation/CatchAllShouldOmitOperationCanceledAnalyzer.cs

```diff
@@ -0,0 +1,136 @@
+    using Particular.Analyzers.Extensions;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class CatchAllShouldOmitOperationCanceledAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            DiagnosticDescriptors.CatchAllShouldOmitOperationCanceled);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                if (!(startContext.Compilation.GetTypeByMetadataName("System.Threading.CancellationToken") is INamedTypeSymbol cancellationTokenType))
+                {
+                    return;
+                }
+
+                startContext.RegisterSyntaxNodeAction(
+                    analyzeContext => Analyze(analyzeContext, cancellationTokenType),
+                    SyntaxKind.TryStatement);
+            });
+        }
+
+        static void Analyze(SyntaxNodeAnalysisContext context, INamedTypeSymbol cancellationTokenType)
+        {
+            if (!(context.Node is TryStatementSyntax tryStatement))
+            {
+                return;
+            }
+
+            foreach (var catchClause in tryStatement.Catches)
+            {
+                var catchType = GetCatchType(catchClause);
+
+                if (catchType == "OperationCanceledException" || catchType == "System.OperationCanceledException")
+                {
+                    return;
+                }
+
+                if (catchType == "Exception" || catchType == "System.Exception")
+                {
+                    if (CatchFiltersOutOperationCanceled(catchClause, context))
+                    {
+                        return;
+                    }
+
+                    var tryBlockCalls = tryStatement.Block.DescendantNodes().OfType<InvocationExpressionSyntax>();
+
+                    foreach (var call in tryBlockCalls)
+                    {
+                        if (call.ArgumentList.Arguments.Any(arg => IsCancellationToken(arg.Expression, context, cancellationTokenType)))
+                        {
+                            context.ReportDiagnostic(DiagnosticDescriptors.CatchAllShouldOmitOperationCanceled, catchClause.CatchKeyword);
+                            return;
+                        }
+                    }
+
+                }
+            }
+        }
+
+        static bool CatchFiltersOutOperationCanceled(CatchClauseSyntax catchClause, SyntaxNodeAnalysisContext context)
+        {
+            if (catchClause.Filter == null)
+            {
+                return false;
+            }
+
+            var logicalNotExpression = catchClause.Filter.FilterExpression.ChildNodes().OfType<PrefixUnaryExpressionSyntax>().FirstOrDefault();
```

Comment:
@DavidBoike apologies, my suggestion here broke this. This fixes it:

```suggestion
            var logicalNotExpression = catchClause.Filter.ChildNodes().OfType<PrefixUnaryExpressionSyntax>().FirstOrDefault();
```